src/nvim/terminal.c

```diff
@@ -0,0 +1,1005 @@
+{
+  event_push((Event) {.handler = on_refresh}, false);
+  refresh_pending = false;
+}
+
+// Refresh all invalidated terminals
+static void on_refresh(Event event)
+{
+  if (exiting) {
+    // bad things can happen if we redraw when exiting, and there's no need to
+    // update the buffer.
+    return;
+  }
+  Terminal *term;
+  void *stub; (void)(stub);
+  // dont process autocommands while updating terminal buffers. JobActivity can
+  // be used act on terminal output.
+  block_autocmds();
+  map_foreach(invalidated_terminals, term, stub, {
+    if (!term->buf) {
+      // destroyed by `close_buffer`. Dont do anything else
+      continue;
+    }
+    WITH_BUFFER(term->buf, {
+      refresh_size(term);
+      refresh_scrollback(term);
+      refresh_screen(term);
+      redraw_buf_later(term->buf, NOT_VALID);
+    });
+    adjust_topline(term);
+  });
+  pmap_clear(ptr_t)(invalidated_terminals);
+  unblock_autocmds();
+  redraw();
+}
+
+static void refresh_size(Terminal *term)
+{
+  if (!term->pending_resize || term->closed) {
+    return;
+  }
+
+  term->pending_resize = false;
+  int width, height;
+  vterm_get_size(term->vt, &height, &width);
+  term->invalid_start = 0;
+  term->invalid_end = height;
+  term->opts.resize_cb((uint16_t)width, (uint16_t)height, term->opts.data);
+}
+
+// Refresh the scrollback of a invalidated terminal
+static void refresh_scrollback(Terminal *term)
+{
+  int width, height;
+  vterm_get_size(term->vt, &height, &width);
+
+  while (term->sb_pending > 0) {
+    // This means that either the window height has decreased or the screen
+    // became full and libvterm had to push all rows up. Convert the first
+    // pending scrollback row into a string and append it just above the visible
+    // section of the buffer
+    if (((int)term->buf->b_ml.ml_line_count - height) >= (int)term->sb_size) {
+      // scrollback full, delete lines at the top
+      ml_delete(1, false);
+      deleted_lines(1, 1);
+    }
+    fetch_row(term, -term->sb_pending, width);
+    int buf_index = (int)term->buf->b_ml.ml_line_count - height;
+    ml_append(buf_index, (uint8_t *)term->textbuf, 0, false);
+    appended_lines(buf_index, 1);
+    term->sb_pending--;
+  }
+
+  // Remove extra lines at the bottom
+  int max_line_count = (int)term->sb_current + height;
+  while (term->buf->b_ml.ml_line_count > max_line_count) {
+    ml_delete(term->buf->b_ml.ml_line_count, false);
+    deleted_lines(term->buf->b_ml.ml_line_count, 1);
+  }
+}
+
+// Refresh the screen(visible part of the buffer when the terminal is
+// focused) of a invalidated terminal
+static void refresh_screen(Terminal *term)
+{
+  int changed = 0;
+  int added = 0;
+  int height;
+  int width;
+  vterm_get_size(term->vt, &height, &width);
+  // Its possible that the terminal height decreased and `term->invalid_end`
+  // doesn't reflect it yet
+  term->invalid_end = MIN(term->invalid_end, height);
+
+  for (int r = term->invalid_start, linenr = row_to_linenr(term, r);
```

Comment:
Is signed overflow still a problem if we don't use the result? Because if `term->invalid_start` is `INT_MAX` then `linenr` wont be used